EventProc/src/OverlayProducer.cxx

```diff
@@ -0,0 +1,329 @@
+      doPoisson_           = parameters.getParameter< int >("doPoisson");
+      timeSigma_           = parameters.getParameter< double >("timeSpread");
+      timeMean_            = parameters.getParameter< double >("timeMean");
+      overlayProcessName_  = parameters.getParameter< std::string >("overlayProcessName");
+      collections_         = parameters.getParameter< std::vector <std::string> >("overlayHitCollections");
+      seed_                = parameters.getParameter< int >("randomSeed");  // TODO: this will use new common deterministic seeding
+      passName_            = "overlay";
+      verbosity_           = parameters.getParameter< int >("verbosity");
+
+
+	  simParticleCollName_ = parameters.getParameter<std::string>( "collection_name" );
+      simParticlePassName_ = parameters.getParameter<std::string>( "pass_name" );
+      overlayPassName_     = parameters.getParameter<std::string>( "overlay_pass_name" );
+
+      if (verbosity_) {
+		 ldmx_log(debug) <<  "[ OverlayProducer ]: Got parameters \n \t overlayFileName = " << overlayFileName_ 
+				  << "\n\t overlayTreeName = " << overlayTreeName_
+				  << "\n\t numberOverlaidInteractions = " << poissonMu_
+				  << "\n\t doPoisson = " << doPoisson_
+				  << "\n\t overlayProcessName = " << overlayProcessName_
+				  << "\n\t overlayHitCollections = " ;
+		for ( const std::string &coll : collections_ )
+		   ldmx_log(debug) <<  coll << "; " ;
+
+		 ldmx_log(debug) <<  "\n\t randomSeed = " << seed_
+				  << "\n\t timeSpread = " << timeSigma_
+				  << "\n\t timeMean = " << timeMean_
+				  << "\n\t passName hardcoded to " << passName_
+				  << "\n\t verbosity = " << verbosity_ ;
+		
+		//		  << "\n\t  = " << 
+      }
+	  
+
+     
+
+        return;
+    }
+
+  void OverlayProducer::produce(Event& event) {
+	//event is the incoming, simulated event/"hard" process
+	//overlayEvent is the overlay producer's own event. 
+	if (verbosity_ > 2) {
+	   ldmx_log(debug) <<  "[ OverlayProducer ]: produce() starts on simulation event " <<  event.getEventHeader().getEventNumber() ;
+	}
+	
+	// sample a poisson distribution or use a deterministic number of overlay events : this is nEvsOverlay
+	int nEvsOverlay = doPoisson_ ? (int) rndm_->Poisson( poissonMu_ ) : (int)poissonMu_ ;  //either sample or take the number at face value every time 
+
+	//TO DO: find a way to make use of this logic while using nextEvent 
+	// find if lastEvent_ + nEvsOverlay > nEvents in the tree as a whole. in that case, reset lastEvent to 0 or (some random small nb)
+	if (lastEvent_ + nEvsOverlay >= nEventsTot_)
+	  lastEvent_ = 0;
+	
+	if (verbosity_ > 2) {
+	   ldmx_log(debug) <<  "[ OverlayProducer ]: will overlay " <<  nEvsOverlay << " events on the simulated one" ;
+	}
+	
+	
+	
+	/*
+	  if (! overlayFile_->nextEvent() ) {
+	  std::cerr << "Couldn't read next event!" ;
+	  return;
+	  }
+	*/
+	
+	//get event wherever nextEvent left us 
+	Event * overlayEvent = overlayFile_->getEvent();
```

Comment:
Your member variable `overlayEvent_` was attached to the `overlayFile_` using `setupEvent` in `onProcessStart`. This means `overlayEvent_` is the object that is updated when you call `nextEvent`, so you don't need to get a reference to the event here.

Reply:
Thanks for catching that! I'll push a fix. Indeed this code still has a few remnants of scrambled iterations since May 😄  I hope it still parses well enough for figuring out what goes awry with the contribs though.